packages/backend-core/src/middleware/passport/sso/sso.ts

```diff
@@ -0,0 +1,163 @@
+import { generateGlobalUserID } from "../../../db"
+import { authError } from "../utils"
+import * as users from "../../../users"
+import * as context from "../../../context"
+import fetch from "node-fetch"
+import { SSOAuthDetails, SSOUser, User } from "@budibase/types"
+
+type SaveUserOpts = {
+  requirePassword?: boolean
+  hashPassword?: boolean
+  currentUserId?: string
+}
+
+export type SaveUserFunction = (
+  user: SSOUser,
+  opts: SaveUserOpts
+) => Promise<any>
+
+/**
+ * Common authentication logic for third parties. e.g. OAuth, OIDC.
+ */
+export async function authenticate(
+  details: SSOAuthDetails,
+  requireLocalAccount: boolean = true,
+  done: any,
+  saveUserFn?: SaveUserFunction
+) {
+  if (!saveUserFn) {
+    throw new Error("Save user function must be provided")
+  }
+  if (!details.provider) {
```

Comment:
Same as above, can this be done removing the undefined option on typing?

Reply:
yep - doesn't need to be there anymore. this is a hangover from when we were pure js - removed in: https://github.com/Budibase/budibase/pull/9672/commits/e7461e180dc1ee447ff23f3b885feca732fb2710